// @sap-devx/yeoman-ui-types/dist/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Prompts = exports.MessageType = exports.Severity = exports.AppWizard = void 0;
var AppWizard = /** @class */ (function () {
    function AppWizard() {
    }
    AppWizard.create = function (genOptions) {
        if (genOptions === void 0) { genOptions = {}; }
        return genOptions.appWizard ? genOptions.appWizard : new EmptyAppWizard();
    };
    return AppWizard;
}());
exports.AppWizard = AppWizard;
var EmptyAppWizard = /** @class */ (function (_super) {
    __extends(EmptyAppWizard, _super);
    function EmptyAppWizard() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    EmptyAppWizard.prototype.showProgress = function (message) { };
    EmptyAppWizard.prototype.showWarning = function (message, type) { };
    EmptyAppWizard.prototype.showError = function (message, type) { };
    EmptyAppWizard.prototype.showInformation = function (message, type) { };
    return EmptyAppWizard;
}(AppWizard));
var Severity;
(function (Severity) {
    Severity[Severity["error"] = 0] = "error";
    Severity[Severity["warning"] = 1] = "warning";
    Severity[Severity["information"] = 2] = "information";
})(Severity = exports.Severity || (exports.Severity = {}));
var MessageType;
(function (MessageType) {
    MessageType[MessageType["prompt"] = 0] = "prompt";
    MessageType[MessageType["notification"] = 1] = "notification";
})(MessageType = exports.MessageType || (exports.MessageType = {}));
var Prompts = /** @class */ (function () {
    function Prompts(items) {
        if (items === void 0) { items = []; }
        this.items = items;
    }
    Prompts.prototype.splice = function (start, deleteCount, items) {
        var _a;
        if (items) {
            items = Array.isArray(items) ? items : [items];
            (_a = this.items).splice.apply(_a, __spreadArrays([start, deleteCount], items));
        }
        else {
            this.items.splice(start, deleteCount);
        }
        if (this.callback) {
            this.callback(this.items);
        }
    };
    Prompts.prototype.setCallback = function (callback) {
        this.callback = callback;
        callback(this.items);
    };
    Prompts.prototype.size = function () {
        return this.items.length;
    };
    return Prompts;
}());
exports.Prompts = Prompts;
