// @sap/generator-hdb-project/generators/app/index.js
'use strict';
const Generator = require( 'yeoman-generator' );
const types = require( '@sap-devx/yeoman-ui-types' );
const Validations = require( '../Validations' );
module.exports = class extends Generator {
    constructor( args, opts ) {
        super( args, opts );

        this.setPromptsCallback = fn => {
            if( this.prompts ) {
                this.prompts.setCallback( fn );
            }
        };

        const prompts = [
            { name: 'Add Basic Information', description: 'Provide basic project information' }
        ];

        this.prompts = new types.Prompts( prompts );

        this.option( 'babel' );
        this.opts = {
            projectInformation: {},
            prompts: this.prompts,
            vscode: opts.vscode
        };
    }

    async prompting() {
        const that = this;
        let prompts = [
            {
                type: 'input',
                name: 'projectName',
                message: 'Project Name',
                validate: function( projectName ){
                    const errorMessage = 'The name must start with a letter and may contain letters, digits, dashes,' +
                        ' periods, and underscores (but no spaces). The periods can only be used in between other ' +
                        'characters (for example, ab.a34.d).';
                    return Validations.isValid( projectName, errorMessage, that );
                }
            }
        ];
        this.basicInformation = await this.prompt( prompts );
        this.mtaInformation = {
            applicationVersion: '0.0.1',
            applicationId: this.basicInformation.projectName
        };
        this.opts.projectInformation.basicInformation = this.basicInformation;
        this.opts.projectInformation.mtaInformation = this.mtaInformation;
    }
    async initializing() {
        this.composeWith( require.resolve( '@sap/generator-add-hdb-module/generators/app' ), this.opts );
    }
};
