// @sap/generator-base-mta-module/generators/app/messages.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = void 0;
exports.messages = {
    MODULE_NAME_VALIDATION_ERR: "The name must start with a letter and may contain letters, digits, dashes, " +
        " periods and underscores (but no spaces). The periods can only be used in between other characters " +
        "(for example: ab.a34.d).",
    EMPTY_MODULE_NAME_VALIDATION_ERR: "You must provide a module name to continue.",
    TARGET_FOLDER_PATH_VALIDATION_ERR: "You must provide the path to an existing folder to continue.",
    MTA_PATH_VALIDATION_ERR: "You must provide the path to the \"mta.yaml\" file to continue.",
    MTA_PATH_PROMPT_MSG: "Provide the path to the project’s \"mta.yaml\" file",
    MTA_PATH_PROMPT_HINT: "You must provide a path to continue.",
    TARGET_FOLDER_PROMPT_MSG: "Enter the path to the folder where the module should be generated",
    MODULE_NAME_PROMPT_MSG: "Enter the module name",
    MODULE_NAME_EXISTS_ERR: "A module with this name already exists in the selected 'mta.yaml' file. You must provide a unique name.",
    MODULE_FOLDER_EXISTS_ERR: "A folder with this name already exists in the target directory. You must provide a unique name.",
    GENERAL_STEP: "Set Basic Properties"
};
